agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package datadog.trace.bootstrap
+
+import io.opentelemetry.auto.bootstrap.InternalJarURLHandler
+import io.opentelemetry.auto.util.test.AgentSpecification
+import spock.lang.Shared
+
+class InternalJarURLHandlerTest extends AgentSpecification {
+
+  @Shared
+  URL testJarLocation = new File("src/test/resources/classloader-test-jar/testjar-jdk8").toURI().toURL()
+
+
+  def "test extract packages"() {
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    expect:
+    packages == handler.getPackages()
+
+    where:
+    dir      | packages
+    "parent" | ['a', 'a.b', 'a.b.c'].toSet()
+    "child"  | ['x', 'x.y', 'x.y.z'].toSet()
+  }
+
+  def "test get URL"() {
+    setup:
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    when:
+    URLConnection connection = handler.openConnection(new URL('file://' + file))
+    assert connection != null
+    byte[] data = new byte[128]
+    int read = connection.getInputStream().read(data)
+    then:
+    read > 0
+
+    where:
+    dir      | file
+    "parent" | '/a/A.class'
+    "parent" | '/a/b/B.class'
+    "parent" | '/a/b/c/C.class'
+    "child"  | '/x/X.class'
+    "child"  | '/x/y/Y.class'
+    "child"  | '/x/y/z/Z.class'
+  }
+
+
+  def "test read class twice"() {
```

Comment:
Don't understand, what does this test do exactly?

Reply:
Still don't understand :( Why we should specifically test that we can read class on our second attempt? How does second attempt differ from the first or the third one?

Reply:
@iNikem I'm going to assume that you are keeping an open mind here so I will try to explain :).  

There are two reasons related to a caching mechanism:

1. **Constrain what is cached:** I wanted to test that a load does not cache the stream after releasing it and losing control over its lifecyle. What's special about the second load which makes it distinct from the first load is that _there was a **predecessor**_.  There is a caching mechanism in place, (the `JarEntry` is cached with the class name), but what if someone comes along later and tried to cache something closeable like the stream instead? That would be unfortunate, but this test would break if that ever happened.
2. **Hitting the branch where caching has occurred**: there happen to be about 800 classloads early on which go all the way down the chain and end up in the leaf-level classloader _twice_. Preventing this from happening would have been a much more invasive change than mitigating it with a cache. Unless you test the load twice, you won't test the branch where the cache is hit. The logic actually dismisses the cache after a hit, based on knowledge of this pattern, making the third load indistinguishable from the first load, hence the lack of a test for it. Good question though.

Does that make sense? :)

Reply:
Great, this explanation does make sense :) What is now missing, is this (or abridged) explanation in the test itself :) For that poor developer who will try to understand this test half a year in the future.

Reply:
thx @richardstartin! @iNikem i will add this explanation to the test 👍

Reply:
@iNikem I would love to live in a world where each and every test I'm unfamiliar with happens to have a nice two paragraph explanation attached and I never need to go and take a look at the system under test, but it's not the kind of luxury I've experienced myself.

Reply:
If our world is not ideal, it does not mean we should not try to make it a better place. I hope you are not saying “There are a lot of bad code out there, why bother writing good one?” :)

Reply:
hm, i don't think that's what @richardstartin was saying. but anyways, i think we're good here. i will add the comment to this test.

Reply:
comment added 👍